README.md
restfn
======

[![Build Status](https://travis-ci.org/MichaelBlume/restfn.png?branch=master)](https://travis-ci.org/MichaelBlume/restfn)
[![Dependencies Status](http://jarkeeper.com/MichaelBlume/restfn/status.png)](http://jarkeeper.com/MichaelBlume/restfn)

This is a library for quickly bolting an admin REST API on to your long-running
app. Specifically it's a library for specifying the URL structure of that API
using nested Clojure literals. You pass a map literal specifying your API
structure to get-rest-handler and it gives you back a handler of the kind you
could then pass to Ring. Restfn doesn't depend on Ring, but it more or less
assumes you do.

You add restfn to your app in more or less the usual way.

In Leiningen:

```clj
[restfn "0.1.1"]
```

In Maven:

```xml
<dependency>
  <groupId>restfn</groupId>
  <artifactId>restfn</artifactId>
  <version>0.1.1</version>
</dependency>
```

Then within your project:

```clj
(ns your.ns
  (:use (restfn core)))
```

This will import one function, get-rest-handler, and one protocol,
RestSerializable. Pass your map to get-rest-handler, and pass the result to
Ring.

How does a map literal represent a URL structure? Well, you can probably
partially guess, but it works a bit like this. You pass in an object
implementing IFn, probably a map. Then the client sends in a URI. The URI is
split into segments. Restfn applies your object to the first segment, then
applies the result to the next segment, until there's no more, then returns the
result.

So, observe:

```clj
(get-rest-handler
  {"map" {"foo" "bar"}})
```

If you hit this handler at **/map**, you'll get the JSON-encoded map
```{"foo" "bar"}```.  If you hit it at **/map/foo** you'll get a JSON-encoding
of the string ```"bar"```.  Super-useful, right?

```clj
(get-rest-handler
  {"stats" {"total" total-stats
            "worker" get-stats-for-worker}})
```

Let's say ```total-stats``` is some stats object for your app, it implements the
```RestSerializable``` protocol:

```clj
(defprotocol RestSerializable
  (rest-serialize [this]
   "Convert to something Cheshire can JSONify"))
```

and returns a map containing stats for your app. Let's say
```get-stats-for-worker``` is a similar function which takes a worker number
and returns stats for that worker (either a map or a similar object
implementing ```RestSerializable```).

Then **/stats/total** will get you that map of total stats, and
**/stats/worker/3** will get you the stats for worker 3 (the "3" is
transparently converted to int before being fed to the function).

```clj
(get-rest-handler
  {"processing" {:delete stop-processing!
                 :post   start-processing!}})
```

This creates one endpoint, **/processing**. If you **POST** to it, it'll start
up your app. If you **DELETE** it, it'll stop your app.

Restfn has a bit of extra smarts for handling java ```Map```s and ```List```s.
By extra smarts, I mean, if it can't apply an object to a segment, it'll try
calling ```.get``` instead.

If your request causes an exception, you'll get a status code 500 and a
traceback.

Functions in a ```{:post fn :delete fn :put fn}``` map can optionally take a
parameter -- that parameter will be the request object. If you wrap your
handler in middleware that enhances the request object, these fns will, of
course, get the enhanced request.


HISTORY.md
0.1.0 introduced a feature where a nil result would produce a 404


test/restfn/test.clj
(ns restfn.test
  (:import (java.util ArrayList))
  (:use (clojure test)
        (clojure [string :only [split-lines]])
        (restfn core)
        (cheshire [core :only [parse-string]])))

(extend-type java.util.regex.Pattern
  RestSerializable
  (rest-serialize [r] #(re-matches r %)))

(def delay-calls (atom 0))

(def handler
  (get-rest-handler
    {"list" [1 3 5]
     "map" {"foo" "bar"}
     "atom" (atom 5)
     "doubleatom" (atom (atom 7))
     "javalist" (ArrayList. [5 9 7])
     "simplefn" inc
     "complexfn" (fn [x] {"key" x})
     "pattern" #"(.*)and(.*)"
     "seq" (for [x (range 3)]
             (reify RestSerializable (rest-serialize [_] x)))
     "rest-delay" (rest-delay (swap! delay-calls inc))
     "rest" {:post (fn [req] (:req-key req))
             :delete (fn [] "deleted")}}))

(defn object-from-request [req]
  (let [res (handler req)]
    (is (= (:status res) 200) res)
    (is (= (:headers res) {"Content-Type" "application/json"}) res)
    (parse-string (:body res))))

(defn object-from-uri [uri]
  (object-from-request {:uri uri :request-method :get}))

(deftest simplegets
  (are [uri result]
    (= result (object-from-uri uri))
    "list" [1 3 5]
    "/list/" [1 3 5]
    "/list/0" 1
    "/list/1" 3
    "atom" 5
    "doubleatom" 7
    "map" {"foo" "bar"}
    "map/foo" "bar"
    "javalist/" [5 9 7]
    "javalist/0" 5
    "simplefn/5" 6
    "complexfn/3" {"key" 3}
    "complexfn/hello/key" "hello"
    "pattern/fooandbar" ["fooandbar" "foo" "bar"]
    "seq" [0 1 2]))

(deftest test-404
  (is (= 404 (:status (handler {:uri "/null" :request-method :get})))))

(deftest testrest
  (are [req result]
    (= result (object-from-request req))
    {:uri            "rest"
     :request-method :post
     :req-key        [1 2 3]} [1 2 3]
    {:uri            "rest"
     :request-method :delete} "deleted"))

(deftest testerror
  (let [result (handler {:uri "list/5" :request-method :get})]
    (is (= (:status result) 500))
    (is (= (:headers result) {"Content-Type" "text/plain"}))
    (is (re-matches #".*IndexOutOfBoundsException.*"
                    ((split-lines (:body result)) 0)))))

(deftest test-delay
  (let [delay-calls-init @delay-calls]
    (doseq [n (range 1 100)]
      (is (= (+ n delay-calls-init)
             (object-from-uri "rest-delay"))))))


src/restfn/core.clj
(ns restfn.core
  (:import (java.io StringWriter PrintWriter)
           (clojure.lang IFn ArityException))
  (:use (clojure [string :only [split]])
        (cheshire [core :only [generate-string]])))

(defn- try-parse-int [s]
  (try (Integer/parseInt s)
    (catch NumberFormatException e s)))

(defprotocol RestSerializable
  (rest-serialize [this]
   "Convert to something Cheshire can JSONify"))

(defmacro rest-delay [& forms]
  `(reify RestSerializable
     (rest-serialize [_] ~@forms)))

(extend Object
  RestSerializable
  {:rest-serialize identity})

(extend-type clojure.lang.IDeref
  RestSerializable
  (rest-serialize [this]
    (rest-serialize @this)))

(extend nil
  RestSerializable
  {:rest-serialize identity})

(defn- descend-one [obj-raw seg-raw]
  (let [obj (rest-serialize obj-raw)
        seg (try-parse-int seg-raw)]
    (if (instance? IFn obj)
      (obj seg)
      (.get obj seg))))

(defn- do-result [raw-result req]
  (let [result (if (seq? raw-result)
                 (map rest-serialize raw-result)
                 (rest-serialize raw-result))]
    (if (map? result)
      (if-let [action-fn (result (:request-method req))]
        (try (action-fn req)
          (catch ArityException _ (action-fn)))
        result)
      result)))

(defn- exc->str [^Throwable e]
  (let [wrtr (StringWriter.)
        pwrtr (PrintWriter. wrtr)]
    (.printStackTrace e pwrtr)
    (str wrtr)))

(defn get-rest-handler [rest-tree]
  (fn [req]
    (try
      (let [uri      (:uri req)
            uri-segs (remove empty? (split uri #"/"))
            result   (reduce descend-one rest-tree uri-segs)
            body-obj (do-result result req)
            body     (str
                       (generate-string body-obj {:pretty true})
                       \newline)]
        {:status (if (nil? body-obj) 404 200)
         :headers {"Content-Type" "application/json"}
         :body body})
      (catch Throwable e
        {:status 500
         :headers {"Content-Type" "text/plain"}
         :body (exc->str e)}))))
